### PR TITLE
add methods to enableParentAssetFallbacks and resolve addresses using…

### PR DESCRIPTION
**Pull request by nandubatchu**
… assetMatcher

### CODE REVIEW

File: src/application/services/curx-address-resolver.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        return userAddress;
+    }
+    public resolveAddressByAssetMatcher = (assetMatcher: IAssetMatcher): IAddress => {
+        log.debug("asset matcher provided:", assetMatcher);
+        let userAddress: IAddress|undefined;
+        if (!assetMatcher.assetIdentifierValue) {
+            userAddress = this.resolveAddressByAssetGroup(assetMatcher.assetGroup);
+        } else {
+            if (!this.userCruxAssetTranslator) {
+                throw new BaseError(null, "user client's assetTranslator is required when assetIdentifierValue is provided");
```
Comment (SanchayPahalwani) [inline src/application/services/curx-address-resolver.ts]: Can be replaced with a proper PackageErrorCode

File: src/infrastructure/services/api-clients.ts [hunk 2]
```diff
@@ -50,10 +50,14 @@ export class GaiaServiceApiClient {
...
-            return Boolean(filename !== UPLOADABLE_JSON_FILES.CLIENT_CONFIG || data.indexOf("BlobNotFound") > 0 || data.indexOf("NoSuchKey") > 0);
-        });
-        return responseBody;
+        try {
+            const responseBody: any = await cachedFunctionCall(cacheStorage, options.url, cacheTTL, httpJSONRequest, [options], async (data) => {
+                return Boolean(filename !== UPLOADABLE_JSON_FILES.CLIENT_CONFIG || data.indexOf("BlobNotFound") > 0 || data.indexOf("NoSuchKey") > 0);
+            });
+            return responseBody;
+        } catch (error) {
+            throw new BaseError(error, "Error while retrieving from Gaia");
```
Comment (SanchayPahalwani) [inline src/infrastructure/services/api-clients.ts]: There is already a PackageErrorCode for these kind of failures which should be used

File: src/application/clients/crux-wallet-client.ts [hunk 3]
```diff
@@ -121,15 +124,34 @@ export class CruxWalletClient {
...
+            cruxUser,
+        });
+        const userAddress = addressResolver.resolveAddressBySymbol(walletCurrencySymbol);
+        return userAddress;
+    }
+
+    @throwCruxClientError
+    public resolveAssetAddressForCruxID = async (fullCruxID: string, assetMatcher: IAssetMatcher): Promise<IAddress> => {
+        await this.initPromise;
+        const tag = "resolving_address";
```
Comment (SanchayPahalwani) [inline src/application/clients/crux-wallet-client.ts]: Can we use different tags to identify the number of users actually using this new feature while resolving?

Comment (SanchayPahalwani) [inline src/application/clients/crux-wallet-client.ts]: Can we use different tags to identify the number of users actually using this new feature while resolving?

File: src/application/services/curx-address-resolver.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import { CruxUser, IAddress } from "../../core/entities/crux-user";
+import { BaseError, ErrorHelper, getLogger, PackageErrorCode } from "../../packages";
+import { CruxAssetTranslator, IAssetMatcher } from "./crux-asset-translator";
+const log = getLogger(__filename);
+export interface ICruxAddressResolverOptions {
+    cruxAssetTranslator: CruxAssetTranslator;
+    cruxUser: CruxUser;
+    userCruxAssetTranslator?: CruxAssetTranslator;
+}
+export class CruxAddressResolver {
```
Comment (quixote911) [inline src/application/services/curx-address-resolver.ts]: We shouldn't need this component at application service level. 
Apart from translation of symbol, we need to move everything inside the user.

The AssetTranslator class also should not need any changes. We are not making any changes to translation process.

File: src/application/clients/crux-wallet-client.ts [hunk 5]
```diff
@@ -133,15 +136,34 @@ export class CruxWalletClient {
         if (!cruxUser) {
             throw ErrorHelper.getPackageError(null, PackageErrorCode.UserDoesNotExist);
         }
-        const assetId = this.cruxAssetTranslator.symbolToAssetId(walletCurrencySymbol);
-        if (!assetId) {
-            throw ErrorHelper.getPackageError(null, PackageErrorCode.AssetIDNotAvailable);
+        const addressResolver = new CruxAddressResolver({
```
Comment (quixote911) [inline src/application/clients/crux-wallet-client.ts]: This function should not change

File: src/application/clients/crux-wallet-client.ts [hunk 6]
```diff
@@ -133,15 +136,34 @@ export class CruxWalletClient {
...
+            cruxAssetTranslator: this.cruxAssetTranslator,
+            cruxUser,
+        });
+        const userAddress = addressResolver.resolveAddressBySymbol(walletCurrencySymbol);
+        return userAddress;
+    }
+
+    @throwCruxClientError
+    public resolveAssetAddressForCruxID = async (fullCruxID: string, assetMatcher: IAssetMatcher): Promise<IAddress> => {
```
Comment (quixote911) [inline src/application/clients/crux-wallet-client.ts]: This is essentially what we need - 

```
@throwCruxClientError
public resolveAssetAddressForCruxID = async (fullCruxID: string, assetMatcher: IAssetMatcher): Promise<IAddress> => {
    await this.initPromise;
    const tag = "resolving_address_with_matcher";
    const cruxUser = await this.getCruxUserByID(fullCruxID.toLowerCase(), tag);
    return cruxUser.getAddressByAssetMatcher(assetMatcher)
}
```
?

File: src/application/clients/crux-wallet-client.ts (hunk 5 above)
Comment (nandubatchu) [reply src/application/clients/crux-wallet-client.ts]: as discussed, not querying the addressMap with asset

File: src/application/clients/crux-wallet-client.ts [hunk 7]
```diff
@@ -174,15 +189,27 @@ export class CruxWalletClient {
         return {success, failures};
     }
 
+    @throwCruxClientError
+    public putParentAssetFallbacks = async (symbolAssetGroups: string[]): Promise<string[]> => {
```
Comment (quixote911) [inline src/application/clients/crux-wallet-client.ts]: 1. Can you add some inline documentation here
2. Naming consistency with `symbolAssetGroups`

**Pull request merged**